### ISSUE TITLE
Eliminate unnecessary "where true=1" for QBE

### ISSUE THREAD

**Issue by quaff** [spring-projects/spring-data-jpa#2282]
`JpaRepository.findAll(Example)` will generate unnecessary `where true=1` if matched predicates is empty, It's possible to eliminate it.
https://github.com/spring-projects/spring-data-jpa/blob/acf01b6c7140a56cd1c3caa67409b4d4f922937d/src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java#L106-L108

**Comment by quaff** [comment]
I create PR #2283
And I have 2 other proposals:
1. check the predicate if it is equal to `cb.isTrue(cb.literal(true))` then return `null`, drawback it is not possible by JPA api, it depends on vender implementation.
https://github.com/spring-projects/spring-data-jpa/blob/acf01b6c7140a56cd1c3caa67409b4d4f922937d/src/main/java/org/springframework/data/jpa/repository/support/SimpleJpaRepository.java#L939-L941
2. return `null` instead of `cb.isTrue(cb.literal(true))`, drawback is it breaks contract
https://github.com/spring-projects/spring-data-jpa/blob/acf01b6c7140a56cd1c3caa67409b4d4f922937d/src/main/java/org/springframework/data/jpa/convert/QueryByExamplePredicateBuilder.java#L106-L108

**Comment by mp911de** [comment]
Thanks for reaching out. Care to elaborate what issue `true=1`is causing for you and why it would make sense to remove it?

**Comment by quaff** [comment]
> Thanks for reaching out. Care to elaborate what issue `true=1`is causing for you and why it would make sense to remove it?

```java
	@GetMapping("/customers")
	public List<Customer> list(Customer example) {
		return customerRepository.findAll(Example.of(example));
	}
```
`example` object will be empty if no parameters present, and spring-data-jpa will append a meaningless `where ?=1` and passing `true`

**Comment by mp911de** [comment]
I agree with aesthetics here, but does `true=1` prevent any functionality or does it break something?

**Comment by quaff** [comment]
> I agree with aesthetics here, but does `true=1` prevent any functionality or does it break something?

It doesn't prevent any functionality, but I think it could harm a little bit performance (not verified).

**Comment by mp911de** [comment]
Typically, databases optimize these conditions away.

**Comment by gregturn** [comment]
@schauder I have drafted a modification that essentially swaps `where true=1` with a `null`. The one test case this broke I fixed. But I'm not quite sure this adequately verifies the change works.

It appears that EclipseLink, if it sees a `null` in a `CriteriaQuery.where()` clause, will simply clear out the `where` clause. No harm there.

But Hibernate 5.5 and 5.6 appear to try and figure out the type of the clause, and since the type safety of this being a `Predicate` appears to get lost along the way, will seemingly wrap it in a faulty type.

```java
  public Predicate wrap(Expression<Boolean> expression) {
    if (Predicate.class.isInstance(expression)) {
      return (Predicate)expression;
    } else {
      return (Predicate)(PathImplementor.class.isInstance(expression) ? new BooleanAssertionPredicate(this, expression, Boolean.TRUE) : new BooleanExpressionPredicate(this, expression));
    }
  }
```

Our `Specification` and `SpecificationComposition` appear to nicely handle `Predicate` being a `null`, as does `SimpleJpaRepository` when applying a `Specification` to fashion a query. I'm a little fuzzy on if this is adequate coverage for what we use it for, and ample warning for end users.
